service/history/replication/ack_manager.go

```diff
@@ -123,32 +124,42 @@ func (p *ackMgrImpl) NotifyNewTasks(
 		return
 	}
 	maxTaskID := tasks[0].GetTaskID()
+	maxVisibilityTimestamp := tasks[0].GetVisibilityTime()
 	for _, task := range tasks {
 		if maxTaskID < task.GetTaskID() {
 			maxTaskID = task.GetTaskID()
 		}
+		if maxVisibilityTimestamp.Before(task.GetVisibilityTime()) {
+			maxVisibilityTimestamp = task.GetVisibilityTime()
+		}
 	}
 
 	p.Lock()
 	defer p.Unlock()
 	if p.maxTaskID == nil || *p.maxTaskID < maxTaskID {
 		p.maxTaskID = &maxTaskID
 	}
+	if p.maxTaskVisibilityTimestamp == nil || p.maxTaskVisibilityTimestamp.Before(maxVisibilityTimestamp) {
+		p.maxTaskVisibilityTimestamp = timestamp.TimePtr(maxVisibilityTimestamp)
+	}
 }
 
-func (p *ackMgrImpl) GetMaxTaskID() int64 {
+func (p *ackMgrImpl) GetMaxTaskInfo() (int64, time.Time) {
```

Comment:
nit: should we define a `struct` for return value?

Reply:
don't think it is necessary here.